File: Исходники/Глава 2. Часть 1/Жадные алгоритмы/4. Задача о заявках/C++/Продуктовый/Applications.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

bool compare(pair<int, int> first_application,
    pair<int, int> second_application) {
    return first_application.second < second_application.second;
}

int count_applications(vector<pair<int, int>>  applications) {
    sort(applications.begin(), applications.end(), compare);
    vector<pair<int, int>> amount_application;

    pair<int, int> last_application = make_pair(0, 0);
    for (pair<int, int> application : applications) {
        if (last_application.second <= application.first) {
            amount_application.push_back(application);
            last_application = application;
        }
    }
    return amount_application.size();
}

File: Исходники/Глава 2. Часть 1/Жадные алгоритмы/4. Задача о заявках/Python/Applications.py
# Даны заявки на проведение занятий в некоторой аудитории.
# В каждой заявке указаны начало и конец занятия (si и fi для i-ой заявки).
# Нужно из всех заявок оставить как можно больше так, чтобы они не пересекались.
# При этом если одна заявка закончилась во время t,
# а следующая началась во время t, то их можно ставить подряд.

applications = [(5, 7), (7, 10), (6, 7), (12, 15), (13, 14)]


def count_applications(applications):
    applications = sorted(applications, key=lambda x: x[1])
    ans = list()

    last_application = (0, 0)
    for i in applications:
        if last_application[1] <= i[0]:
            ans.append(i)
            last_application = i
    return len(ans)


# print(count_applications(applications))